L86-aspect-based-python-bazel-rules.md

```diff
@@ -0,0 +1,50 @@
+Aspect-based Python Bazel Rules
+----
+* Author(s): Michael Beardsworth
+* Approver: gnossen
+* Status: Draft
+* Implemented in: Bazel and Starlark
+* Last updated: Sept. 8, 2021
+* Discussion at: https://groups.google.com/g/grpc-io/c/NXdtW5xe9Js
+
+## Abstract
+
+Proposes modifications to gRPC's Bazel rules for Python codegen to improve
+dependency tracking.
+
+## Background
+
+* gRPC provides build rule logic written in Starlark to support Protobuf and gRPC Python codegen with Bazel.
+* These rules (`py_proto_library` and `py_grpc_library`) each generate code from a single `proto_library`'s protos.
+* The produced Python libraries (or equivalently-functioning PyInfo providers) do not express dependencies on the generated Python code for the `proto_library`'s dependencies.
+* Users are surprised by this behavior. See [#23769](https://github.com/grpc/grpc/issues/23769) for an example bug.
+* This behavior is particularly surprising for Google-internal users. The internal versions of `py_proto_library` and `py_grpc_library` propagate Python dependencies.
+
+### Related Proposals:
+`py_proto_library` and `py_grpc_library` are implemented in [python_rules.bzl](https://github.com/grpc/grpc/blob/bazel/python_rules.bzl). There doesn't seem to be a proposal for that, though.
+
+## Proposal
+
+* We propose rebuilding the `py_proto_library` and `py_grpc_library` rules to use [aspects](https://docs.bazel.build/versions/main/skylark/aspects.html) to perform code generation steps.
+  This approach corresponds to the Google-internal design for these rules.
+* The `_gen_py_proto_aspect` aspect visits `proto_library` rules to generate Python code for Protobuf.
```

Comment:
We might want to note the protobuf Python code is still generated by us instead of `proto_library`.